File: id/src/main/java/com/simprints/id/tools/LanguageHelper.java
package com.simprints.id.tools;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;

import java.util.Locale;

// TODO: improve language switching. Either don't go against android, or implement a bulletproof solution.
public class LanguageHelper {

    public static Configuration selectLanguage(String languageString) {
        Locale locale = localeFor(languageString);

        Locale.setDefault(locale);
        Configuration config = new Configuration();
        config.setLocale(locale);

        return config;
    }

    public static Locale localeFor(String languageString) {
        Locale locale;

        /*
         * Languages are usually two  or three characters, except if they also contain a region.
         * If they contain a region they follow the format [language]-r[REGION] e.g. fa-rAF
         */
        if (languageString.contains("-r") && languageString.length() > 2) {

            int indexOfFlag = -1;
            for (int i = 0; i < languageString.length() - 1; i++) {
                if (languageString.substring(i, i + 2).equals("-r")) {
                    indexOfFlag = i;
                }
            }

            String language = languageString.substring(0, indexOfFlag);
            String country = languageString.substring(indexOfFlag + 1, languageString.length());

            locale = new Locale(language, country);

        } else {
            locale = new Locale(languageString);
        }

        return locale;
    }

    public static void setLanguage(Context context, String languageString) {
        Resources res = context.getResources();
        res.updateConfiguration(LanguageHelper.selectLanguage(languageString), res.getDisplayMetrics());
    }
}

File: id/src/main/java/com/simprints/id/tools/LanguageHelper.kt
package com.simprints.id.tools

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import com.simprints.id.data.prefs.PreferencesManagerImpl
import com.simprints.id.data.prefs.settings.SettingsPreferencesManagerImpl.Companion.LANGUAGE_DEFAULT
import com.simprints.id.data.prefs.settings.SettingsPreferencesManagerImpl.Companion.LANGUAGE_KEY
import timber.log.Timber
import java.util.*

object LanguageHelper {

    lateinit var prefs: SharedPreferences
    var language: String
        get() {
            return prefs.getString(LANGUAGE_KEY, LANGUAGE_DEFAULT)!!
        }
        set(value) {
            prefs.edit().putString(LANGUAGE_KEY, value).apply()
        }

    fun init(ctx: Context){
        prefs = ctx.getSharedPreferences(PreferencesManagerImpl.PREF_FILE_NAME, Context.MODE_PRIVATE)
    }

    private fun configurationWithSpecificLocale(languageString: String): Configuration =
        Configuration().apply {
            val locale = localeFor(languageString)
            Locale.setDefault(locale)
            setLocale(locale)
        }


    /*
     * Languages are usually two  or three characters, except if they also contain a region.
     * If they contain a region they follow the format [language]-r[REGION] e.g. fa-rAF
     */
    private fun localeFor(languageString: String): Locale {
        val localeParts = languageString.split("-r")
        val language = localeParts[0]
        return if (localeParts.size > 1) {
            Locale.Builder()
                .setLanguage(language)
                .setRegion(localeParts[1]).build()

        } else {
            Locale(localeParts[0])
        }
    }

    fun contextWithSpecificLanguage(context: Context, languageString: String): Context =
        context.createConfigurationContext(configurationWithSpecificLocale(languageString))

    fun getLanguageConfigurationContext(ctx: Context): Context {
        println("Testing language: $language")
        val conf = configurationWithSpecificLocale(language)
        return ctx.createConfigurationContext(conf)
    }
}
